# Problems:

### Problem 1: Gender Recognition by Voice

From the description file at https://data.world/ml-research/gender-recognition-by-voice:

In order to analyze gender by voice and speech, a training database was required. A database was built using thousands of samples of male and female voices, each labeled by their gender of male or female. Voice samples were collected from the following resources:

*  [The Harvard-Haskins Database of Regularly-Timed Speech](http://nsi.wegall.net/)
*  Telecommunications & Signal Processing Laboratory (TSP) Speech Database at McGill University
*  [VoxForge Speech Corpus](http://www.repository.voxforge1.org/downloads/SpeechCorpus/Trunk/Audio/Main/8kHz_16bit/)
*  [Festvox CMU_ARCTIC Speech Database at Carnegie Mellon University](http://festvox.org/cmu_arctic/dbs_awb.html)

Each voice sample is stored as a .WAV file, which is then pre-processed for acoustic analysis using the specan function from the WarbleR R package. Specan measures 22 acoustic parameters on acoustic signals for which the start and end times are provided.

The output from the pre-processed WAV files were saved into a CSV file, containing 3168 rows and 21 columns (20 columns for each feature and one label column for the classification of male or female). You can download the pre-processed dataset in CSV format, using the link above
Acoustic Properties Measured

The following acoustic properties of each voice are measured:

*    __duration:__ length of signal
*    __meanfreq:__ mean frequency (in kHz)
*    __sd:__ standard deviation of frequency
*    __median:__ median frequency (in kHz)
*    __Q25:__ first quantile (in kHz)
*    __Q75:__ third quantile (in kHz)
*    __IQR:__ interquantile range (in kHz)
*    __skew:__ skewness (see note in specprop description)
*    __kurt:__ kurtosis (see note in specprop description)
*    __sp.ent:__ spectral entropy
*    __sfm:__ spectral flatness
*    __mode:__ mode frequency
*    __centroid:__ frequency centroid (see specprop)
*    __peakf:__ peak frequency (frequency with highest energy)
*    __meanfun:__ average of fundamental frequency measured across acoustic signal
*    __minfun:__ minimum fundamental frequency measured across acoustic signal
*    __maxfun:__ maximum fundamental frequency measured across acoustic signal
*    __meandom:__ average of dominant frequency measured across acoustic signal
*    __mindom:__ minimum of dominant frequency measured across acoustic signal
*    __maxdom:__ maximum of dominant frequency measured across acoustic signal
*    __dfrange:__ range of dominant frequency measured across acoustic signal
*    __modindx:__ modulation index. Calculated as the accumulated absolute difference between adjacent measurements of fundamental frequencies divided by the frequency range

The gender of the speaker is given in the __label__ column. 

Note, the features for duration and peak frequency (peakf) were removed from training. Duration refers to the length of the recording, which for training, is cut off at 20 seconds. Peakf was omitted from calculation due to time and CPU constraints in calculating the value. In this case, all records will have the same value for duration (20) and peak frequency (0).

Load file using the code below. 

#### Question 1:

Which two features are most indicative of gendered voice?

#### Question 2:

Preform Linear Regression, Logistic Regression, and Quadratic Discriminant Analysis on the features, graphing the resulting fits. How does the two feature fit compare to the fit on all features?

In [ ]:
import pandas as pd

data = pd.read_csv("https://raw.githubusercontent.com/tipthederiver/Math-7243-2020/master/Datasets/GenderedVoice/voice.csv")

data.head()

## Problem 2: MRI Data

The dementia level for the Oasis 1 MRI dataset is based on a patient assessment. As a result, it is not clear whether the levels of 0, .5, 1 and 2 should actually be understood as meaningfully numeric, or if they in fact are categorical labels. 

In this problem we want to treat them as categorical. However, we would also like to construct a slightly larger dataset, as we have seen that for images our 700 may not be sufficient. To construct a larger dataset we will again down sample the images, however this time we will use the down sampling to expand the dataset instead of throwing data away. After fixing a down sample rate $D$, we will construct one image out of the pixels $nD$, for $n = 1,2,\ldots, $. We will also construct $n D+i$, for $i = 1,\ldots, D$. This way, by down sampling with a rate $D$, we construct $D$ more pictures. 

__Note:__ It is very import that we perform the train test split _before_ we expand the dataset through down sampling. If not, we are effectively training on the test data. 

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import matplotlib

file_dir = 'C:/PATH_TO_IMAGE_DIRECTORY/'
file_dir = 'C:/Users/Admin/Downloads/Student Data/MRILargeSlices/'

labels = pd.read_csv(file_dir + 'labels.csv')
display(labels)
y = labels.CDR

Unnamed: 0             ID M/F Hand  Age  Educ  SES  MMSE  CDR  eTIV  \
0             0  OAS1_0001_MR1   F    R   74   2.0  3.0  29.0  0.0  1344   
1             1  OAS1_0001_MR1   F    R   74   2.0  3.0  29.0  0.0  1344   
2             2  OAS1_0001_MR1   F    R   74   2.0  3.0  29.0  0.0  1344   
3             3  OAS1_0002_MR1   F    R   55   4.0  1.0  29.0  0.0  1147   
4             4  OAS1_0002_MR1   F    R   55   4.0  1.0  29.0  0.0  1147   
..          ...            ...  ..  ...  ...   ...  ...   ...  ...   ...   
697         697  OAS1_0456_MR1   M    R   61   5.0  2.0  30.0  0.0  1637   
698         698  OAS1_0456_MR1   M    R   61   5.0  2.0  30.0  0.0  1637   
699         699  OAS1_0457_MR1   F    R   62   3.0  3.0  26.0  0.0  1372   
700         700  OAS1_0457_MR1   F    R   62   3.0  3.0  26.0  0.0  1372   
701         701  OAS1_0457_MR1   F    R   62   3.0  3.0  26.0  0.0  1372   

      nWBV    ASF  Delay                                       Filename  
0    0.743  1.306    NaN   OAS1_0001_MR1_mpr_n4_anon_111_t88_gfc_44.png  
1    0.743  1.306    NaN   OAS1_0001_MR1_mpr_n4_anon_111_t88_gfc_86.png  
2    0.743  1.306    NaN  OAS1_0001_MR1_mpr_n4_anon_111_t88_gfc_128.png  
3    0.810  1.531    NaN   OAS1_0002_MR1_mpr_n4_anon_111_t88_gfc_44.png  
4    0.810  1.531    NaN   OAS1_0002_MR1_mpr_n4_anon_111_t88_gfc_86.png  
..     ...    ...    ...                                            ...  
697  0.780  1.072    NaN   OAS1_0456_MR1_mpr_n4_anon_111_t88_gfc_86.png  
698  0.780  1.072    NaN  OAS1_0456_MR1_mpr_n4_anon_111_t88_gfc_128.png  
699  0.766  1.279    NaN   OAS1_0457_MR1_mpr_n3_anon_111_t88_gfc_44.png  
700  0.766  1.279    NaN   OAS1_0457_MR1_mpr_n3_anon_111_t88_gfc_86.png  
701  0.766  1.279    NaN  OAS1_0457_MR1_mpr_n3_anon_111_t88_gfc_128.png  

[702 rows x 14 columns]

In [3]:
data = np.zeros([702, 36608])

for n, file_name in enumerate(labels.Filename):
    data[n,:] = np.mean(matplotlib.image.imread(file_dir + file_name),axis=2).reshape(-1)

    
    
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(data, y, test_size=0.2, random_state=0)
print(y_train.shape, y_test.shape)

(561,) (141,)


We want to sample the data array using the `data[start:stop:step]` slice paradigm. This means we are taking elements of the array `data` starting at `start`, ending at `stop` with step `step`. This is why previously `data[::DS]` down sampled at a rate of DS. For example, 

    lst = list(range(165)); lst[6::10]
    
returns

    [6, 16, 26, 36, 46, 56, 66, 76, 86, 96, 106, 116, 126, 136, 146, 156]

We need to create two new arrays, one of shape $[561\times DS, 36608/DS]$ containing the down sampled data, and one of shape $[561\times DS]$ containing the labels. The for each of the $N_{train}$ images in the training array, we need to create $DS$ new down sampled images, with the downsample starting from $i$:

`Xds_train[n+i, :] = X_train[i::DS]`

This will split our images into DS down sampled images. We then need to be sure to save out the appropriate label:

In [39]:
DS = 8             # Downsample rate, must be a multiple of 36608

N_train = y_train.shape[0]  # The length of the training data
y_train = np.array(y_train)

if 36608/DS % 1 > 0:
    print("Downsample rate is not a multiple of 36608")
    DS = 1
    im_size = 36608
else:
    im_size = int(36608/DS)

Xds_train = np.zeros([N_train*DS, im_size])
yds_train = np.zeros([N_train*DS])
   
for n in range(N_train):
    for i in range(DS):
        Xds_train[n*DS+i,:] = X_train[n,i::DS]
        yds_train[n*DS+i] = y_train[n]

In [44]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

qda = LinearDiscriminantAnalysis(store_covariance=True)
qda.fit(Xds_train, (2*yds_train).astype("int"))
print("Score: %.3f"%qda.score(Xds_train,(2*yds_train).astype("int")))

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:388: UserWarning: Variables are collinear.
  warnings.warn("Variables are collinear.")


Score: 1.000


In [42]:
(2*yds_train).astype("int")

array([0, 0, 0, ..., 2, 2, 2])

### Question 1:

Based on the code above, downsample the test data in the same way. 

### Question 2:

Perform LDA, QDA, Logistic Regression and Categorical Linear Regression on the down sampled Oasis 1 dataset. How do these compare to linear regression?